Revise partial_bindless release note (#24506)
bevyengine/bevy: _release-content/release-notes/partial_bindless_metal.md

---
title: Partial Bindless and Reduced Bind Group Overhead
authors: ["@holg"]
pull_requests: [23436]
---

In an ideal world, Bevy users could write a single application and ship it everywhere, with every last one of the messy cross-platform differences beautifully abstracted away.

> That can be a bit hard though.

Reviewer: could probably omit this sentence, it doesn't really do anything imo

> However, users were reporting that rendering complex scenes on Mac and iOS was markedly slower than on comparable hardware.

Reviewer 2: It’s not clear what the comparison is between. Should this say “comparable non-Mac hardware”?

Author: ```suggestion
However, users were reporting that rendering complex scenes on Mac and iOS was markedly slower than on comparable non-Apple hardware.
```

> Looking into it, the lack of bindless rendering support was to blame.

Reviewer: these are fine, but feels a little handwavey? like "we found" and "looking into it" just feel wishywashy haha

it would come across stronger if there was more concrete background like "In previous versions of Bevy user reports and benchmarks of complex scenes on mac and iOS showed much worse performance than other platforms. After investigation of the graphics pipeline used we found that it wasn't taking advantage of bindless rendering where it could." (continues into line 12)

Bindless rendering is how modern engines handle scenes with many different materials efficiently: shaders index into shared pools of textures and buffers rather than rebinding them per draw call.

> Both Metal (Apple's GPU API) and DX12 (an older Windows API) have partial bindless support:

Reviewer 3: dx12 is also a modern api

```suggestion
Both Metal (Apple's GPU API) and DX12 (a Windows graphics API) have partial bindless support:
```

they permit texture binding arrays but not buffer binding arrays.

> Historically, Bevy required both to enable bindless, which excluded Metal entirely, even for materials that never use buffer arrays.

Reviewer 2: ```suggestion
Historically, Bevy required support for both features before it would use bindless, which excluded Metal entirely, even for materials that never use buffer arrays.
```

Most materials, including `StandardMaterial`, do not need buffer array support.
To ensure those materials take the fast path, Bevy now checks the actual needs of each material.
If you only need texture arrays, your material can be rendered efficiently across Bevy's desktop platforms.

> If you use `#[uniform(..., binding_array(...))]`, expect unusually poor performance when using Metal or DX12.

Reviewer 3: I feel "poor performance" is a bit too harsh.

We've also fixed two important correctness bugs in the process.

> First, we discovered that the sampler limit check was testing the wrong metric: `max_samplers_per_shader_stage` counts binding slots, but the relevant limit is `max_binding_array_sampler_elements_per_shader_stage`, the array element count (a mismatch that could silently exceed hardware limits).

Reviewer 3: Seems `max_binding_array_sampler_elements_per_shader_stage` is larger than `max_samplers_per_shader_stage` so it's unlikely to exceed.

```suggestion
First, we discovered that the sampler limit check was testing the wrong metric: `max_samplers_per_shader_stage` counts binding slots, but the relevant limit is `max_binding_array_sampler_elements_per_shader_stage`, the array element count (a mismatch that could incorrectly disable bindless).
```

Second, Bevy now also skips creating binding array slots for resource types a material doesn't use, staying within Metal's hard 31 argument buffer slot limit and reducing overhead on all platforms.

Benchmarked on Bistro Exterior (698 materials), 5-minute runs:

| GPU                      | Avg FPS improvement | Min FPS improvement | Memory      |
| ------------------------ | ------------------- | ------------------- | ----------- |
| Apple M2 Max (Metal)     | +18%                | +77%                | −57 MB RAM  |
| NVIDIA 5060 Ti           | +84%                | +174%               | Same        |
| AMD Vega 8 / Ryzen 4800U | Same                | Same                | −88 MB VRAM |
| Intel i360P              | +15%                | Same                | Same        |
| Intel Iris XE            | Same                | Same                | Same        |

[Bistro] is a demanding, fairly realistic scene.
While bindless limitations remain frustrating, especially on Mac where Vulkan isn't an option,

> it's lovely to see those performance gains, and to know that Bevy itself is no longer artifically holding our users back.

Author: ```suggestion
it's lovely to see those performance gains, and to know that Bevy itself is no longer artificially holding our users back.
```

Second, Bevy now also skips creating binding array slots for resource types a material doesn't use, staying within Metal's hard 31 argument buffer slot limit and reducing overhead on all platforms.

Benchmarked on Bistro Exterior (698 materials), 5-minute runs:

| GPU                      | Avg FPS improvement | Min FPS improvement | Memory      |
| ------------------------ | ------------------- | ------------------- | ----------- |
| Apple M2 Max (Metal)     | +18%                | +77%                | −57 MB RAM  |
| NVIDIA 5060 Ti           | +84%                | +174%               | Same        |
| AMD Vega 8 / Ryzen 4800U | Same                | Same                | −88 MB VRAM |
| Intel i360P              | +15%                | Same                | Same        |
| Intel Iris XE            | Same                | Same                | Same        |

[Bistro] is a demanding, fairly realistic scene.
While Metal's bindless limitations remain frustrating,
it's lovely to see those performance gains, and to know that Bevy is not artificially holding performance on iOS and macOS back.

> [Bistro]: https://developer.nvidia.com/orca/amazon-lumberyard-bistro

Author: ```suggestion
[Bistro]: https://developer.nvidia.com/orca/amazon-lumberyard-bistro

```
